client.stream.Position.toBytes fails to serialize -1

**Problem description**

Calling `client.stream.Position.toBytes` fails with some events returned by ReadEvents. Events consist of short UTF-8 strings such as "123456789". I believe this problem occurs for streams with multiple segments in which 1 segment has no events.

Stack trace is below.
```
Caused by: java.lang.IllegalArgumentException: writeCompactLong can only serialize longs in the interval [0, 2^62), given -1.
	at io.pravega.common.io.serialization.RevisionDataOutputStream.writeCompactLong(RevisionDataOutputStream.java:104)
	at io.pravega.client.stream.impl.PositionImpl$PositionSerializer.lambda$write00$2(PositionImpl.java:138)
	at io.pravega.common.io.serialization.RevisionDataOutputStream.writeMap(RevisionDataOutputStream.java:343)
	at io.pravega.client.stream.impl.PositionImpl$PositionSerializer.write00(PositionImpl.java:137)
	at io.pravega.common.io.serialization.VersionedSerializer$SingleType.serializeContents(VersionedSerializer.java:336)
	at io.pravega.common.io.serialization.VersionedSerializer$SingleType.serialize(VersionedSerializer.java:304)
	at io.pravega.common.io.serialization.VersionedSerializer$WithBuilder.serialize(VersionedSerializer.java:511)
	at io.pravega.common.io.serialization.VersionedSerializer.serialize(VersionedSerializer.java:130)
	at io.pravega.client.stream.impl.PositionImpl.toBytes(PositionImpl.java:174)
	at io.pravega.example.gateway.grpc.PravegaServerImpl.readEvents(PravegaServerImpl.java:204)
	... 9 common frames omitted
```

Below shows the first event in the stream. position.getBytes returns this exception. The description field is from position.toString() which shows a -1 for one of the segments.
```
event: "1234"
 position {
   description: "examples/memory-eater-in/1.#epoch.0=-1, examples/memory-eater-in/0.#epoch.0=12"
 }
 event_pointer {
   bytes: "\000\000\001\000\000\000\000\'\000#examples/memory-eater-in/0.#epoch.0\000\014"
   description: "examples/memory-eater-in/0.#epoch.0:0-12"
 }
 stream_cut {
   text: "H4sIAAAAAAAAADOwSq1IzC3ISS3Wz03NzS+q1E1NLEkt0s3MszLQMQRiAytDIx1dQwCG5bJbKAAAAA=="
   description: "examples/memory-eater-in:0=12, 1=-1"
 }
```

**Problem location**
Pravega Client 0.7.0

**Suggestions for an improvement**
The serializer must use signed integers.

--- a/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
@@ -118,25 +118,34 @@ public class PositionImpl extends PositionInternal {
 
         @Override
         protected byte getWriteVersion() {
-            return 1;
+            return 0;
         }
 
         @Override
         protected void declareVersions() {
             version(0).revision(0, this::write00, this::read00)
-                      .revision(1, this::write01, this::read01);
-            version(1).revision(0, this::write10, this::read10);
+                      .revision(1, this::write01, this::read01)
+                      .revision(2, this::write02, this::read02);
         }
 
         private void read00(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
-            Map<Segment, Long> map = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactLong);
+            // RevisionDataInput::readCompactSignedLong can read data written using RevisionDataOutput.writeCompactLong
+            Map<Segment, Long> map = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactSignedLong);
             builder.ownedSegments(map);
         }
 
         private void write00(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+            // This is used to ensure serialized data can be read by the older client.
+            // Note: the older client does not serialize if the offset has -1L as value.
             Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
-                                        (out, offset) -> out.writeCompactLong(offset));
+                                        (out, offset) -> {
+                                            if ( offset < 0) {
+                                                out.writeCompactSignedLong(offset);
+                                            } else {
+                                                out.writeCompactLong(offset);
+                                            }
+                                        });
         }
         
         private void read01(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
@@ -149,17 +158,17 @@ public class PositionImpl extends PositionInternal {
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializer::writeRange);
         }
 
-        private void read10(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
-            Map<Segment, Long> ownedSegments = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactSignedLong);
-            builder.ownedSegments(ownedSegments);
-            read01(revisionDataInput, builder);
-        }
-
-        private void write10(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+        private void write02(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+            // serialize offset segment again as CompactSignedLong ensure serialization
             Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
                     (out, offset) -> out.writeCompactSignedLong(offset));
-            write01(position, revisionDataOutput);
+        }
+
+        // This method is invoked only if the data is serialized with revision as 2.
+        private void read02(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
+            Map<Segment, Long> map = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactSignedLong);
+            builder.ownedSegments(map);
         }
 
         private static void writeRange(RevisionDataOutput out, Range range) throws IOException {
@@ -188,7 +197,7 @@ public class PositionImpl extends PositionInternal {
         ByteArraySegment serialized = SERIALIZER.serialize(this);
         return ByteBuffer.wrap(serialized.array(), serialized.arrayOffset(), serialized.getLength());
     }
-    
+
     @SneakyThrows(IOException.class)
     public static Position fromBytes(ByteBuffer buff) {
         return SERIALIZER.deserialize(new ByteArraySegment(buff));

--- a/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
@@ -53,7 +53,7 @@ public class PositionImpl extends PositionInternal {
     }
     
     @Builder(builderClassName = "PositionBuilder")
-    private PositionImpl(Map<Segment, Long> ownedSegments, Map<Segment, Range> segmentRanges) {
+    PositionImpl(Map<Segment, Long> ownedSegments, Map<Segment, Range> segmentRanges) {
         this.ownedSegments = ownedSegments;
         if (segmentRanges == null) {
             this.segmentRanges = Collections.emptyMap();
@@ -118,13 +118,14 @@ public class PositionImpl extends PositionInternal {
 
         @Override
         protected byte getWriteVersion() {
-            return 0;
+            return 1;
         }
 
         @Override
         protected void declareVersions() {
             version(0).revision(0, this::write00, this::read00)
                       .revision(1, this::write01, this::read01);
+            version(1).revision(0, this::write10, this::read10);
         }
 
         private void read00(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
@@ -147,7 +148,20 @@ public class PositionImpl extends PositionInternal {
             Map<Segment, Range> map = position.segmentRanges;
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializer::writeRange);
         }
-        
+
+        private void read10(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
+            Map<Segment, Long> ownedSegments = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactSignedLong);
+            builder.ownedSegments(ownedSegments);
+            read01(revisionDataInput, builder);
+        }
+
+        private void write10(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+            Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
+            revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
+                    (out, offset) -> out.writeCompactSignedLong(offset));
+            write01(position, revisionDataOutput);
+        }
+
         private static void writeRange(RevisionDataOutput out, Range range) throws IOException {
             double low, high;
             if (range == null) {

--- a/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
@@ -124,8 +124,7 @@ public class PositionImpl extends PositionInternal {
         @Override
         protected void declareVersions() {
             version(0).revision(0, this::write00, this::read00)
-                      .revision(1, this::write01, this::read01)
-                      .revision(2, this::write02, this::read02);
+                      .revision(1, this::write01, this::read01);
         }
 
         private void read00(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
@@ -135,7 +134,6 @@ public class PositionImpl extends PositionInternal {
         }
 
         private void write00(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
-            // This is used to ensure serialized data can be read by the older client.
             // Note: the older client does not serialize if the offset has -1L as value.
             Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
@@ -156,19 +154,6 @@ public class PositionImpl extends PositionInternal {
         private void write01(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
             Map<Segment, Range> map = position.segmentRanges;
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializer::writeRange);
-        }
-
-        private void write02(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
-            // serialize offset segment again as CompactSignedLong ensure serialization
-            Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
-            revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
-                    RevisionDataOutput::writeCompactSignedLong);
-        }
-
-        // This method is invoked only if the data is serialized with revision as 2.
-        private void read02(RevisionDataInput revisionDataInput, PositionBuilder builder) throws IOException {
-            Map<Segment, Long> map = revisionDataInput.readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactSignedLong);
-            builder.ownedSegments(map);
         }
 
         private static void writeRange(RevisionDataOutput out, Range range) throws IOException {

--- a/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java
@@ -162,7 +162,7 @@ public class PositionImpl extends PositionInternal {
             // serialize offset segment again as CompactSignedLong ensure serialization
             Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
-                    (out, offset) -> out.writeCompactSignedLong(offset));
+                    RevisionDataOutput::writeCompactSignedLong);
         }
 
         // This method is invoked only if the data is serialized with revision as 2.

--- a/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
@@ -26,6 +26,9 @@ import java.util.Random;
 import java.util.concurrent.Callable;
 import java.util.stream.Collectors;
 
+import io.pravega.common.io.serialization.RevisionDataInput;
+import io.pravega.common.io.serialization.RevisionDataOutput;
+import io.pravega.common.io.serialization.VersionedSerializer;
 import lombok.val;
 import org.apache.commons.io.output.ByteArrayOutputStream;
 import org.apache.commons.lang3.RandomStringUtils;
@@ -72,10 +75,81 @@ public class SerializationTest {
     public void testPosition() {
         PositionImpl pos = new PositionImpl(ImmutableMap.of(new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0, 1), 2L));
         ByteBuffer bytes = pos.toBytes();
-        Position pos2 = Position.fromBytes(bytes);
-        assertEquals(pos, pos2);
+        Position pos1 = Position.fromBytes(bytes);
+        assertEquals(pos, pos1);
     }
-    
+
+    @Test
+    public void testPositionNegative() {
+        PositionImpl pos = new PositionImpl(ImmutableMap.of(new SegmentWithRange(Segment.fromScopedName("foo/bar/0"), 0, 0.5), 2L,
+                new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0.5, 1), 1L));
+        ByteBuffer bytes = pos.toBytes();
+        Position pos1 = Position.fromBytes(bytes);
+        assertEquals(pos, pos1);
+    }
+
+    @Test
+    public void testPostionImplCompatabilityV0() throws Exception {
+        PositionImpl pos = new PositionImpl(ImmutableMap
+                .of(new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0, 1), 2L));
+
+        // Obtain version 0 serialized data
+        final byte[] bufV0 = new PositionSerializerV0().serialize(pos).array();
+        // deserialize it using current version 1 serialization and ensure compatibility.
+        assertEquals(pos, Position.fromBytes(ByteBuffer.wrap(bufV0)));
+    }
+
+    private static class PositionSerializerV0 extends VersionedSerializer.Direct<PositionImpl> {
+
+        @Override
+        protected byte getWriteVersion() {
+            return 0;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(0).revision(0, this::write00, this::read00)
+                      .revision(1, this::write01, this::read00);
+        }
+
+        private void write00(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+            Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
+            revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
+                    (out, offset) -> out.writeCompactLong(offset));
+        }
+
+        private void write01(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
+            Map<Segment, SegmentWithRange.Range> map = position.getOwnedSegmentRangesWithOffsets().keySet().stream()
+                                                               .collect(Collectors
+                                                                       .toMap(SegmentWithRange::getSegment, SegmentWithRange::getRange));
+            revisionDataOutput
+                    .writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializerV0::writeRange);
+        }
+
+        private void read00(RevisionDataInput revisionDataInput, PositionImpl target) throws IOException {
+            //NOP
+        }
+
+        private static void writeRange(RevisionDataOutput out, SegmentWithRange.Range range) throws IOException {
+            double low, high;
+            if (range == null) {
+                low = -1;
+                high = -1;
+            } else {
+                low = range.getLow();
+                high = range.getHigh();
+            }
+            out.writeDouble(low);
+            out.writeDouble(high);
+        }
+
+        private static SegmentWithRange.Range readRange(RevisionDataInput in) throws IOException {
+            double low = in.readDouble();
+            double high = in.readDouble();
+            return (low < 0 || high < 0) ? null : new SegmentWithRange.Range(low, high);
+        }
+    }
+
     @Test
     public void testStreamCut() {
         StreamCutImpl cut = new StreamCutImpl(Stream.of("Foo/Bar"), ImmutableMap.of(Segment.fromScopedName("Foo/Bar/1"), 3L));

--- a/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
@@ -88,7 +88,7 @@ public class SerializationTest {
 
         // Serialize via the old serialization logic.
         // Note: the older serialization logic does not work with -1L as offset.
-        final byte[] bufOld = new PositionSerializerR1().serialize(new PositionR1(pos)).array();
+        final byte[] bufOld = new PositionSerializerOld().serialize(new PositionOld(pos)).array();
         // deserialize it using latest revision and ensure compatibility.
         assertEquals(pos, Position.fromBytes(ByteBuffer.wrap(bufOld)));
     }
@@ -101,21 +101,21 @@ public class SerializationTest {
 
         // deserialize it using the old serialization logic and ensure compatibility.
         // Note: The old serialization logic does not allow users to serialize / deserialize with -1L as offset.
-        PositionSerializerR1 oldSerializer = new PositionSerializerR1();
+        PositionSerializerOld oldSerializer = new PositionSerializerOld();
 
-        PositionR1 pos1 = new PositionR1(null, null);
+        PositionOld pos1 = new PositionOld(null, null);
         oldSerializer.deserialize(new ByteArraySegment(newBuf.array()), pos1);
         assertEquals(pos, pos1.getPostionImpl());
     }
 
     // Mutable class to validate older serialization.
     @AllArgsConstructor
-    private class PositionR1 {
+    private class PositionOld {
 
         private Map<Segment, Long> ownedSegments;
         private Map<Segment, SegmentWithRange.Range> segmentRanges;
 
-        public PositionR1(PositionImpl impl) {
+        public PositionOld(PositionImpl impl) {
             this.ownedSegments = impl.getOwnedSegmentsWithOffsets();
             this.segmentRanges = impl.getOwnedSegmentRangesWithOffsets().keySet().stream()
                                      .collect(Collectors
@@ -128,7 +128,7 @@ public class SerializationTest {
     }
 
     // Serializer to simiulate revision 1 of serialization.
-    private static class PositionSerializerR1 extends VersionedSerializer.Direct<PositionR1> {
+    private static class PositionSerializerOld extends VersionedSerializer.Direct<PositionOld> {
 
         @Override
         protected byte getWriteVersion() {
@@ -142,28 +142,28 @@ public class SerializationTest {
                       .revision(1, this::write01, this::read01);
         }
 
-        private void write00(PositionR1 position, RevisionDataOutput revisionDataOutput) throws IOException {
+        private void write00(PositionOld position, RevisionDataOutput revisionDataOutput) throws IOException {
             Map<Segment, Long> map = position.ownedSegments;
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
                     (out, offset) -> out.writeCompactLong(offset));
 
         }
 
-        private void write01(PositionR1 position, RevisionDataOutput revisionDataOutput) throws IOException {
+        private void write01(PositionOld position, RevisionDataOutput revisionDataOutput) throws IOException {
             Map<Segment, SegmentWithRange.Range> m1 = position.segmentRanges;
             revisionDataOutput
-                    .writeMap(m1, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializerR1::writeRange);
+                    .writeMap(m1, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializerOld::writeRange);
         }
 
-        private void read00(RevisionDataInput revisionDataInput, PositionR1 target) throws IOException {
+        private void read00(RevisionDataInput revisionDataInput, PositionOld target) throws IOException {
             Map<Segment, Long> map = revisionDataInput
                     .readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactLong);
             target.ownedSegments = map;
         }
 
-        private void read01(RevisionDataInput revisionDataInput, PositionR1 target) throws IOException {
+        private void read01(RevisionDataInput revisionDataInput, PositionOld target) throws IOException {
             Map<Segment, SegmentWithRange.Range> map1 = revisionDataInput
-                    .readMap(in -> Segment.fromScopedName(in.readUTF()), PositionSerializerR1::readRange);
+                    .readMap(in -> Segment.fromScopedName(in.readUTF()), PositionSerializerOld::readRange);
             target.segmentRanges = map1;
         }
 

--- a/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/SerializationTest.java
@@ -29,6 +29,7 @@ import java.util.stream.Collectors;
 import io.pravega.common.io.serialization.RevisionDataInput;
 import io.pravega.common.io.serialization.RevisionDataOutput;
 import io.pravega.common.io.serialization.VersionedSerializer;
+import lombok.AllArgsConstructor;
 import lombok.val;
 import org.apache.commons.io.output.ByteArrayOutputStream;
 import org.apache.commons.lang3.RandomStringUtils;
@@ -73,33 +74,61 @@ public class SerializationTest {
 
     @Test
     public void testPosition() {
-        PositionImpl pos = new PositionImpl(ImmutableMap.of(new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0, 1), 2L));
-        ByteBuffer bytes = pos.toBytes();
-        Position pos1 = Position.fromBytes(bytes);
-        assertEquals(pos, pos1);
-    }
-
-    @Test
-    public void testPositionNegative() {
         PositionImpl pos = new PositionImpl(ImmutableMap.of(new SegmentWithRange(Segment.fromScopedName("foo/bar/0"), 0, 0.5), 2L,
-                new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0.5, 1), 1L));
+                new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0.5, 1.0), -1L));
         ByteBuffer bytes = pos.toBytes();
         Position pos1 = Position.fromBytes(bytes);
         assertEquals(pos, pos1);
     }
 
     @Test
-    public void testPostionImplCompatabilityV0() throws Exception {
+    public void testPositionImplBackwardCompatibility() throws Exception {
         PositionImpl pos = new PositionImpl(ImmutableMap
                 .of(new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0, 1), 2L));
 
-        // Obtain version 0 serialized data
-        final byte[] bufV0 = new PositionSerializerV0().serialize(pos).array();
-        // deserialize it using current version 1 serialization and ensure compatibility.
-        assertEquals(pos, Position.fromBytes(ByteBuffer.wrap(bufV0)));
+        // Serialize via the old serialization logic.
+        // Note: the older serialization logic does not work with -1L as offset.
+        final byte[] bufOld = new PositionSerializerR1().serialize(new PositionR1(pos)).array();
+        // deserialize it using latest revision and ensure compatibility.
+        assertEquals(pos, Position.fromBytes(ByteBuffer.wrap(bufOld)));
     }
 
-    private static class PositionSerializerV0 extends VersionedSerializer.Direct<PositionImpl> {
+    @Test
+    public void testPositionImplForwardCompatibility() throws Exception {
+        PositionImpl pos = new PositionImpl(ImmutableMap
+                .of(new SegmentWithRange(Segment.fromScopedName("foo/bar/1"), 0, 1), 2L));
+        ByteBuffer newBuf = pos.toBytes();
+
+        // deserialize it using the old serialization logic and ensure compatibility.
+        // Note: The old serialization logic does not allow users to serialize / deserialize with -1L as offset.
+        PositionSerializerR1 oldSerializer = new PositionSerializerR1();
+
+        PositionR1 pos1 = new PositionR1(null, null);
+        oldSerializer.deserialize(new ByteArraySegment(newBuf.array()), pos1);
+        assertEquals(pos, pos1.getPostionImpl());
+    }
+
+    // Mutable class to validate older serialization.
+    @AllArgsConstructor
+    private class PositionR1 {
+
+        private Map<Segment, Long> ownedSegments;
+        private Map<Segment, SegmentWithRange.Range> segmentRanges;
+
+        public PositionR1(PositionImpl impl) {
+            this.ownedSegments = impl.getOwnedSegmentsWithOffsets();
+            this.segmentRanges = impl.getOwnedSegmentRangesWithOffsets().keySet().stream()
+                                     .collect(Collectors
+                                             .toMap(SegmentWithRange::getSegment, SegmentWithRange::getRange));
+        }
+
+        public PositionImpl getPostionImpl() {
+            return new PositionImpl(ownedSegments, segmentRanges);
+        }
+    }
+
+    // Serializer to simiulate revision 1 of serialization.
+    private static class PositionSerializerR1 extends VersionedSerializer.Direct<PositionR1> {
 
         @Override
         protected byte getWriteVersion() {
@@ -108,26 +137,34 @@ public class SerializationTest {
 
         @Override
         protected void declareVersions() {
+
             version(0).revision(0, this::write00, this::read00)
-                      .revision(1, this::write01, this::read00);
+                      .revision(1, this::write01, this::read01);
         }
 
-        private void write00(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
-            Map<Segment, Long> map = position.getOwnedSegmentsWithOffsets();
+        private void write00(PositionR1 position, RevisionDataOutput revisionDataOutput) throws IOException {
+            Map<Segment, Long> map = position.ownedSegments;
             revisionDataOutput.writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()),
                     (out, offset) -> out.writeCompactLong(offset));
+
         }
 
-        private void write01(PositionImpl position, RevisionDataOutput revisionDataOutput) throws IOException {
-            Map<Segment, SegmentWithRange.Range> map = position.getOwnedSegmentRangesWithOffsets().keySet().stream()
-                                                               .collect(Collectors
-                                                                       .toMap(SegmentWithRange::getSegment, SegmentWithRange::getRange));
+        private void write01(PositionR1 position, RevisionDataOutput revisionDataOutput) throws IOException {
+            Map<Segment, SegmentWithRange.Range> m1 = position.segmentRanges;
             revisionDataOutput
-                    .writeMap(map, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializerV0::writeRange);
+                    .writeMap(m1, (out, s) -> out.writeUTF(s.getScopedName()), PositionSerializerR1::writeRange);
         }
 
-        private void read00(RevisionDataInput revisionDataInput, PositionImpl target) throws IOException {
-            //NOP
+        private void read00(RevisionDataInput revisionDataInput, PositionR1 target) throws IOException {
+            Map<Segment, Long> map = revisionDataInput
+                    .readMap(in -> Segment.fromScopedName(in.readUTF()), RevisionDataInput::readCompactLong);
+            target.ownedSegments = map;
+        }
+
+        private void read01(RevisionDataInput revisionDataInput, PositionR1 target) throws IOException {
+            Map<Segment, SegmentWithRange.Range> map1 = revisionDataInput
+                    .readMap(in -> Segment.fromScopedName(in.readUTF()), PositionSerializerR1::readRange);
+            target.segmentRanges = map1;
         }
 
         private static void writeRange(RevisionDataOutput out, SegmentWithRange.Range range) throws IOException {

